src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -306,7 +306,10 @@ ReplicatedRecoveryBackend::recover_delete(
       }
       return seastar::make_ready_future<>();
     }).then_interruptible([this, soid, &stat_diff] {
-      pg.get_recovery_handler()->on_global_recover(soid, stat_diff, true);
+      const auto &missing = pg.get_peering_state().get_pg_log().get_missing();
+      if (!missing.is_missing(soid)) {
+        pg.get_recovery_handler()->on_global_recover(soid, stat_diff, true);
```

Comment:
> the primary tries to recover the delete, but as the missing_loc is cleared, it recognize it as needs to be pulled while there's no missing loc for the object

With this patch, the primary won't clear missing_loc after the replicas finished recovery. Later on, when the primary will try to recover the delete it will be able to successfully pull the missing obj, right?
Just to understand better, why would `!missing.is_missing(soid)` be true after the replicas finish the on delete recovery?

Reply:
> nit: "completed recovery, missing {}"

Will add it, sorry for the delay

Reply:
> This can be a good opportunity to cleanup `bool is_delete` from on_global_recover() since it's no longer used

Yeah

Reply:
> > the primary tries to recover the delete, but as the missing_loc is cleared, it recognize it as needs to be pulled while there's no missing loc for the object
> 
> With this patch, the primary won't clear missing_loc after the replicas finished recovery. Later on, when the primary will try to recover the delete it will be able to successfully pull the missing obj, right?

Yes, you're right.

> Just to understand better, why would `!missing.is_missing(soid)` be true after the replicas finish the on delete recovery?

If the primary has already deleted the object of soid by the time replicas finish the delete recovery, `!missing.is_missing(soid)` will be true.